File: haskell/src/Fib.hs
module Fib where

fib :: Int -> Int
fib 0 = 0
fib 1 = 1
fib n = fib (n - 1) + fib (n - 2)

fib2 :: Int -> Integer
fib2 n = go n (0, 1) where
  go 0 (a, _) = a
  go n (a, b) = go (n - 1) (b, a + b)

fib3 :: Int -> Integer
fib3 n = fibs !! n where
  fibs = 0 : 1 : next fibs where
    next (a : t@(b:_)) = (a + b) : next t

File: elixir/lib/fib.ex
defmodule Fib do
  @moduledoc false

  def fib(0), do: 0
  def fib(1), do: 1
  def fib(n), do: fib(n - 1) + fib(n - 2)

  def fib2(n), do: fib2p(n, {0, 1})
  defp fib2p(0, {a, _}), do: a
  defp fib2p(n, {a, b}), do: fib2p(n - 1, {b, a + b})

  def fib3(n), do: fibs() |> Enum.at(n)
  defp fibs(), do: Stream.unfold({0, 1}, fn {a, b} -> {a, {b, a + b}} end)
end
